File: sdk/golang/tests/e2e/rest/margin_test/margin_debit_test.go
package margin_test

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/extension/interceptor"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/api"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/margin/debit"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"os"
	"testing"
)

var debitApi debit.DebitAPI

func init() {
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	httpOptionBuilder := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		AddInterceptors(interceptor.NewLoggingInterceptor(false, defaultLogger))

	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOptionBuilder.Build()).
		Build()

	client := api.NewClient(option)

	debitApi = client.RestService().GetMarginService().GetDebitAPI()
}

func TestDebitGetBorrowHistoryReq(t *testing.T) {
	// GetBorrowHistory
	// Get Borrow History
	// /api/v3/margin/borrow

	builder := debit.NewGetBorrowHistoryReqBuilder()
	builder.SetCurrency("USDT").SetIsIsolated(true).SetSymbol("BTC-USDT")
	req := builder.Build()

	resp, err := debitApi.GetBorrowHistory(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestDebitBorrowReq(t *testing.T) {
	// Borrow
	// Borrow
	// /api/v3/margin/borrow

	builder := debit.NewBorrowReqBuilder()
	builder.SetCurrency("USDT").SetSize(10.0).SetTimeInForce("IOC").
		SetSymbol("BTC-USDT").SetIsIsolated(true).SetIsHf(true)
	req := builder.Build()

	resp, err := debitApi.Borrow(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestDebitGetInterestHistoryReq(t *testing.T) {
	// GetInterestHistory
	// Get Interest History
	// /api/v3/margin/interest

	builder := debit.NewGetInterestHistoryReqBuilder()
	builder.SetSymbol("BTC-USDT").SetIsIsolated(true)
	req := builder.Build()

	resp, err := debitApi.GetInterestHistory(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestDebitGetRepayHistoryReq(t *testing.T) {
	// GetRepayHistory
	// Get Repay History
	// /api/v3/margin/repay

	builder := debit.NewGetRepayHistoryReqBuilder()
	builder.SetCurrency("USDT").SetIsIsolated(true).SetSymbol("BTC-USDT")
	req := builder.Build()

	resp, err := debitApi.GetRepayHistory(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestDebitRepayReq(t *testing.T) {
	// Repay
	// Repay
	// /api/v3/margin/repay

	builder := debit.NewRepayReqBuilder()
	builder.SetCurrency("USDT").SetSize(10.0).SetSymbol("BTC-USDT").SetIsIsolated(true).SetIsHf(true)
	req := builder.Build()

	resp, err := debitApi.Repay(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

func TestDebitModifyLeverageReq(t *testing.T) {
	// ModifyLeverage
	// Modify Leverage
	// /api/v3/position/update-user-leverage

	builder := debit.NewModifyLeverageReqBuilder()
	builder.SetSymbol("BTC-USDT").SetIsIsolated(true).SetLeverage("3.1")
	req := builder.Build()

	resp, err := debitApi.ModifyLeverage(req, context.TODO())
	if err != nil {
		panic(err)
	}
	data, err := json.Marshal(resp.ToMap())
	if err != nil {
		panic(err)
	}
	fmt.Println("code:", resp.CommonResponse.Code)
	fmt.Println("message:", resp.CommonResponse.Message)
	fmt.Println("data:", string(data))
}

File: sdk/python/test/e2e/rest/margin_test/margin_debit_test.py
import os
import unittest

from kucoin_universal_sdk.api.client import DefaultClient
from kucoin_universal_sdk.extension.interceptor.logging import LoggingInterceptor
from kucoin_universal_sdk.generate.margin.debit.model_borrow_req import BorrowReqBuilder, BorrowReq
from kucoin_universal_sdk.generate.margin.debit.model_get_borrow_history_req import GetBorrowHistoryReqBuilder
from kucoin_universal_sdk.generate.margin.debit.model_get_interest_history_req import GetInterestHistoryReqBuilder
from kucoin_universal_sdk.generate.margin.debit.model_get_repay_history_req import GetRepayHistoryReqBuilder
from kucoin_universal_sdk.generate.margin.debit.model_modify_leverage_req import ModifyLeverageReqBuilder
from kucoin_universal_sdk.generate.margin.debit.model_repay_req import RepayReqBuilder
from kucoin_universal_sdk.model.client_option import ClientOptionBuilder
from kucoin_universal_sdk.model.constants import GLOBAL_API_ENDPOINT, GLOBAL_BROKER_API_ENDPOINT, \
    GLOBAL_FUTURES_API_ENDPOINT
from kucoin_universal_sdk.model.transport_option import TransportOptionBuilder


class MarginApiTest(unittest.TestCase):
    def setUp(self):
        key = os.getenv("API_KEY")
        secret = os.getenv("API_SECRET")
        passphrase = os.getenv("API_PASSPHRASE")

        # create http transport option
        http_transport_option = (
            TransportOptionBuilder()
            .set_interceptors([LoggingInterceptor()])
            .build()
        )

        # create client option
        client_option = (
            ClientOptionBuilder()
            .set_key(key)
            .set_secret(secret)
            .set_passphrase(passphrase)
            .set_spot_endpoint(GLOBAL_API_ENDPOINT)
            .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
            .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
            .set_transport_option(http_transport_option)
            .build()
        )

        # create API client
        client = DefaultClient(client_option)
        kucoin_rest_api = client.rest_service()
        self.api = kucoin_rest_api.get_margin_service().get_debit_api()

    def test_get_borrow_history_req(self):
        """
            get_borrow_history
            Get Borrow History
            /api/v3/margin/borrow
        """

        builder = GetBorrowHistoryReqBuilder()
        builder.set_currency("USDT").set_is_isolated(True).set_symbol("BTC-USDT")
        req = builder.build()
        try:
            resp = self.api.get_borrow_history(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_borrow_req(self):
        """
            borrow
            Borrow
            /api/v3/margin/borrow
        """

        builder = BorrowReqBuilder()
        builder.set_currency("USDT").set_size(10.0).set_time_in_force(BorrowReq.TimeInForceEnum.IOC).set_symbol(
            "BTC-USDT").set_is_isolated(True).set_is_hf(True)
        req = builder.build()
        try:
            resp = self.api.borrow(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_interest_history_req(self):
        """
            get_interest_history
            Get Interest History
            /api/v3/margin/interest
        """

        builder = GetInterestHistoryReqBuilder()
        builder.set_currency("USDT").set_is_isolated(True).set_symbol("BTC-USDT")
        req = builder.build()
        try:
            resp = self.api.get_interest_history(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_get_repay_history_req(self):
        """
            get_repay_history
            Get Repay History
            /api/v3/margin/repay
        """

        builder = GetRepayHistoryReqBuilder()
        builder.set_currency("USDT").set_is_isolated(True).set_symbol("BTC-USDT").set_order_no(
            "67467fcefc3d5000075c5ec2")
        req = builder.build()
        try:
            resp = self.api.get_repay_history(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_repay_req(self):
        """
            repay
            Repay
            /api/v3/margin/repay
        """

        builder = RepayReqBuilder()
        builder.set_currency("USDT").set_size(10).set_symbol("BTC-USDT").set_is_isolated(True).set_is_hf(True)
        req = builder.build()
        try:
            resp = self.api.repay(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e

    def test_modify_leverage_req(self):
        """
            modify_leverage
            Modify Leverage
            /api/v3/position/update-user-leverage
        """

        builder = ModifyLeverageReqBuilder()
        builder.set_symbol("BTC-USDT").set_is_isolated(True).set_leverage("3.1")
        req = builder.build()
        try:
            resp = self.api.modify_leverage(req)
            print("code: ", resp.common_response.code)
            print("msg: ", resp.common_response.message)
            print("data: ", resp.to_json())
        except Exception as e:
            print("error: ", e)
            raise e
